modules/mod-deep-learning/DeepModelManager.cpp

```diff
@@ -0,0 +1,554 @@
+{
+   return GetRootURL(repoID)+filePath;
+}
+
+void DeepModelManager::FetchLocalCards(CardFetchedCallback onCardFetched)
+{
+   FilePaths pathList;
+   FilePaths modelFiles; 
+
+   // NOTE: maybe we should support multiple search paths? 
+   pathList.push_back(DeepModel::DLModelsDir());
+
+   FileNames::FindFilesInPathList(wxT("model.pt"), pathList, 
+                                 modelFiles, wxDIR_FILES | wxDIR_DIRS);
+
+   for (const auto &modelFile : modelFiles)
+   {
+      wxFileName modelPath(modelFile);
+      wxFileName cardPath(modelPath.GetFullPath());
+
+      cardPath.SetFullName("metadata.json");
+
+      if (cardPath.FileExists() && modelPath.FileExists())
+      {
+         ModelCardHolder card = std::make_shared<ModelCard>();
+         bool success = NewCardFromLocal(card, audacity::ToUTF8(cardPath.GetFullPath()));
+         onCardFetched(success, card);
+      }
+   }
+}
+
+void DeepModelManager::FetchRepos(RepoListFetchedCallback onReposFetched)
+{
+   // NOTE: the url below asks for all repos in huggingface
+   // that contain the tag "audacity". 
+   // however, it might be better for us to keep a curated list of 
+   // models which we show to the user, and allow the user to explore huggingface
+   // on their own for more repos
+   // std::string query = mAPIEndpoint + "models?filter=audacity";
+   std::string query = GetFileURL("hugggof/audacity-models", "models.json");
+
+   // TODO: handle exception in main thread. try using a broken url?
+   CompletionHandler handler = 
+   [query, onReposFetched = std::move(onReposFetched)]
+   (int httpCode, std::string body)
+   {
+      RepoIDList repos;
+      if (httpCode != 200)
+      {
+         wxLogError("GET request failed for url %s. Error code: %d", 
+                     query, httpCode);
+         onReposFetched(false, repos); 
+      }
+      else
+      {
+         bool success = true;
+
+         Doc reposDoc;
+         try
+         {
+            reposDoc = parsers::ParseString(body);
+         }
+         catch (const InvalidModelCardDocument &e)
+         {
+            wxLogError("error parsing JSON reponse for fetching repos");
+            wxLogError(e.what());
+            success = false;
+         }
+
+         if (success && reposDoc.IsArray())
+         {
+            for (auto itr = reposDoc.Begin(); itr != reposDoc.End(); ++itr)
+            {
+               wxLogDebug("Found repo with name %s", itr->GetString());
+               repos.emplace_back(itr->GetString());
+            }
+         }
+         else 
+            success = false;
+
+         onReposFetched(success, repos);
+      }
+   };
+
+   doGet(query, handler);
+}
+
+void DeepModelManager::FetchCard(const std::string &repoID, CardFetchedCallback onCardFetched)
+{ 
+   const std::string modelCardUrl = GetFileURL(repoID, "metadata.json");
+   wxLogMessage("Fetching model card from %s", modelCardUrl);
+
+   // TODO: how do you handle an exception inside a thread, like this one? 
```

Comment:
I see this question, and I think a solution might be: https://en.cppreference.com/w/cpp/thread/promise and the related `std::future`, which establish a channel between threads that can perform a one-time communication of either a result value or an exception.  You can read details, while I try to understand your code better to see whether this might really apply here.

Reply:
using wxTheApp->CallAfter works!